[Ready]Removes lag. [MC update]

MrsTonedOne opened the pull request:
This setup allows subsystems to pause themselves until next mc fire if they are about to go over a tick, and resume on the next mc tick
It also makes it so we prioritize running subsystems we skipped because there wasn't enough time in the current tick to run them based on their avg tick usage. (as well as subsystems paused mid way thru)

Adds a macro for adding this to non-mc procs. just drop it in the loop. There is a define version as well for if checks if you want to know if you just slept to avoid lag.

This is all inside #if DM_VERSION defines, so once i test that 509 still works to compile/test this should be good to merge.

SubSystems using the new system:
mob
machine
air
garbage
minimaps (also made a config option that defaults to off to help coders test quicker)
object
lighting
diseases
events
smooth_icon
timer

The rest aren't in the new system because they aren't laggy enough to warrant it.

Procs given the ghetto version of this:
explosion()
singularity/eat()
(These were both changed to use a spiral range to make it look sexy propagating outwards from the center)
del-all
sdql2
mass var edit
maploader

12,24,48 explosion, no lag.

13k active turfs, no lag

super matter sing, no lag

all three at the same time, no lag.

:cl:
rscdel: Lag has been removed from the following systems: mob life()/virus process, machine/object/event processing, atmos, deletion subsystem, lighting, explosions, singularity's eat(), timers, wall smoothing, mass var edit, mass proc calls(and all of sdql2), map loader, and the away mission loader.  
rscdel: Removed limit on bomb cap of 32,64,128, it can now go to the MOON.  
tweak: Minimap generation is now a config option that defaults to off so coders don't have to wait for it to generate to test their code, server operators should note that they need to enable it on production servers that want minimap generation.  
wip: if you find any things that still lag, please report them to MrStonedOne for lag removal.  
experiment: I'm gonna take this time to shill out that, Lummox JR, byond's only dev, coded the tool that made this change possible, he lives off of byond membership donations, if you like the lag removal, become a byond member or at least throw 5 bucks byond's way for doing this.  
experiment: Feel free to ask for bomb cap removals when solo antag, but remember, the later in the round it is, the more likely you'll get it.  
/:cl:
# port notes:

Please port #16245 as well.

This pr was coded to the tune of https://www.youtube.com/watch?v=F0YYoo6oFoU

MrsTonedOne committed:
Removes lag.

This setup allows subsystems to pause themselves until next mc fire if they are about to go over a tick, and resume on the next mc tick
It also makes it so we prioritize running subsystems we skipped because there wasn't enough time in the current tick to run them based on their avg tick usage. (as well as subsystems paused mid way thru)

Adds a macro for adding this to non-mc procs. just drop it in the loop. There is a define version as well for if checks if you want to know if you just slept to avoid lag.

This is all inside #if DM_VERSION defines, so once i test that 509 still works to compile/test this should be good to merge

MMMiracles commented:
i want to believe

MrStonedOne commented:
byond://75.172.96.169:1337 believe

MrsTonedOne committed:
forgot to save the dme

MrStonedOne commented:
actually I'm gonna do a test merge now.

MMMiracles commented:
I BELIEVE

RemieRichards commented:
Lighting looks hella ugly when moving around fast as a ghost.
But its not really possible to move that fast in game normally so I guess it's ok.

Iamgoofball commented:
LAG IS DEAD

Buggy123 commented:
RIP.

PKPenguin321 commented:
this is kickass

optimumtact commented:
Doesn't really remove the lag though, just spreads it across several ticks

MrStonedOne commented:
no, it removes the lag, as lag can only happen if byond misses a tick, and this prevents that.

optimumtact commented:
well no, the lag now will be a subsystem noticeably firing over multiple ticks instead of one

DavidEdwards commented:
If I understand correctly, it is simply transitioning lag from the server/player and pushing it onto the subsystems. So if there is a problem causing atmos to spike, atmos will simply update much slower than everything else.

The player will move around as if there were no lag. This is arguably more important.

optimumtact commented:
I know, just let me be pedantic here David

Incoming5643 commented:
it's kinda bad watching the tesla take 20-30 seconds between shocks

Iamgoofball commented:
We had 8 teslas inside containment and we didn't notice that.

Incoming5643 commented:
Well then there was something weird going on last round. Air keeps freaking out  every so often for some reason.

RemieRichards commented:
@Incoming5643 yeah uh there WAS something wrong last round, MSO found an edge case that causes atmos to be ran over and over.

MrsTonedOne committed:
fixes deadlocking

MrsTonedOne committed:
Adds a way to slow down the mc via var edits under new system, for testing.

JJRcop commented:
Yo yo yo, MC ticker in the house. Let's turn up the beat!

Core0verload commented:
LAG IS DEAD AND FREEZING HAS NO MEANING

Cheridan commented on code/controllers/subsystem/mobs.dm:
```diff
@@ -14,10 +15,16 @@ var/datum/subsystem/mobs/SSmob
+	while(currentrun.len)
+		var/mob/M = currentrun[1]
+		currentrun.Cut(1, 2)
+		if(M)
```
missed a space here :)

Fox-McCloud commented on code/__HELPERS/unsorted.dm:
```diff
@@ -1291,3 +1291,7 @@ proc/pick_closest_path(value)
+
+/proc/pause1tick()
+	. = 1
+	sleep world.tick_lag
```
`sleep(world.tick_lag)`

Fox-McCloud commented:
Question: should subsystems even dynamically adjust their speed anymore with this change? I mean, if you can just defer to the next tick, it seems a bit redundant.

MrStonedOne commented:
in this case dynamic wait will more so mean that its next fire is determined on when it finished processing, not when it started processing.

And subsystems with a low wait could still use dynamic wait to ensure they don't hog the mc's attention too much. lighting, garbage, and air are all 5ds subsystems, if all 3 need 3 mc ticks to do their things then the mc would start to fall behind on running other subsystem.

and mind you, dynamic wait scales with how busy the mc is totally, and world.cpu's current value.

MrsTonedOne committed:
Explosions and singulo use spiral range (and other changes)

Miauw62 commented:
H A B B E D I N G
A
B
B
E
D
I
N
G

KorPhaeron commented:
You are going to need to update line 191 of mine_items.dm since you are removing trange.

I have no idea why travis isn't catching this, but trying to test merge this/update won't compile because of it

@MrStonedOne

Fox-McCloud commented:
Is `trange` really slower than `spiral_range`?

MrStonedOne commented:
kor, this is/was currently test merged, it must be something merged recently

RemieRichards commented:
@Fox-McCloud 
"(These were both changed to use a spiral range to make it look sexy propagating outwards from the center)"

its purely for looks.

Fox-McCloud commented:
I wondered; otherwise, it just starts deleting things from the bottom left to the top right.

KorPhaeron commented:
@MrStonedOne 

The lava thing got merged yeah

MrsTonedOne committed:
Adds lag checks to massedit, sdql2, del-all, and the map loader

MrStonedOne commented:
bonus points, removing lag from the map loader will actually make it much faster.

MrsTonedOne committed:
Adds defines

MrsTonedOne committed:
Removes cap on bomb cap

MrsTonedOne committed:
Lag check on initialization (and on a missed del-all proc)

MrsTonedOne committed:
Adds lag check to diseases and events (and removes lighting from dwait)
lighting removed from dewait because lag checks are better for it and lag checked subsystems that are also dwait systems are given lower priority, we don't want that to happen to lighting.

MrsTonedOne committed:
dynamic wait subsystems will not priority run in any situation, even when paused.

Lumipharon commented:
Neat.

MrsTonedOne committed:
When waking up, we check if there is enough tick left to bother, if not, we sleep for double.

MrsTonedOne committed:
Config option for generating minimaps, defaults to disabled, for coder's sanity

MrsTonedOne committed:
Throwing uses lag check to avoid eating all the cpu

MrsTonedOne committed:
Singulo processes by tile, not atom, to avoid pulling something that's moved.

MrsTonedOne committed:
Adds lag checks to SSsmooth

MrsTonedOne committed:
Fixes the singularity not eating/pulling turfs.

MrsTonedOne committed:
Makes the code compile under 509

PKPenguin321 commented:
Woo

duncathan commented:
The lag is dead. Long live the lag!

Iamgoofball commented:
so should we be using spiral_range and spiral_range_turfs over range() and trange() now?

Stickymayhem commented:
Woah does this mean gas can expand in a pretty circle and look more liquidy

DOES THIS MEAN WE CAN HAVE LIQUIDS
